### PR TITLE
[FLINK-20856][table-planner-blink] Separate the implementation of stream window aggregate nodes

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Separate the implementation of stream window aggregate nodes, including 
StreamExecGroupWindowAggregate
StreamExecGroupWindowTableAggregate
StreamExecPythonGroupWindowAggregate*

## Brief change log

  - *Introduce StreamPhysicalGroupWindowAggregate & StreamPhysicalGroupWindowTableAggregate, and make StreamExecGroupWindowAggregate only extended from ExecNode*
  - *Introduce StreamPhysicalPythonGroupWindowAggregate, and make StreamExecPythonGroupWindowAggregate only extended from ExecNode*

## Verifying this change

This change is a refactoring rework covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+    private final WindowEmitStrategy emitStrategy;
+    private final boolean needRetraction;
+    private final String aggType;
+
+    public StreamExecGroupWindowAggregate(
+            int[] grouping,
+            AggregateCall[] aggCalls,
+            LogicalWindow window,
+            PlannerNamedWindowProperty[] namedProperties,
+            int inputTimeFieldIndex,
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java:100]: this seems not necessary, we can get from LogicalWindow?

**Review by wenlong88 — COMMENTED** [review]
the overall change LGTM, just left a minor comment

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+            String description) {
+        super(Collections.singletonList(inputEdge), outputType, description);
+        this.grouping = grouping;
+        this.aggCalls = aggCalls;
+        this.window = window;
+        this.namedProperties = namedProperties;
+        this.inputTimeFieldIndex = inputTimeFieldIndex;
+        this.emitStrategy = emitStrategy;
+        this.needRetraction = needRetraction;
+        this.aggType = aggType;
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java:115]: I think `aggType` is a too generic name. It seems only be used for printing. Is it possible to eliminate it? Maybe we don't need to mention this is a TableAggregate or Aggregate, the important information is that this is a window.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+            } else if (isRowtimeAttribute(timeField)) {
+                builder = builder.session(toDuration(gap)).withEventTime(timeFieldIndex);
+            } else {
+                throw new UnsupportedOperationException("This should not happen.");
+            }
+        } else {
+            throw new TableException("Unsupported window: " + window.toString());
+        }
+
+        if (emitStrategy.produceUpdates()) {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java:346]: The original implementation of `StreamExecGroupWindowTableAggregate` throws exception if `emitStrategy` produces updates. However, this check is removed now? 

Maybe we can throw this exception in `WindowOperatorBuilder` to keep this class unchanged.

Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java:346]: No, the code snippet is in `StreamPhysicalGroupWindowTableAggregate`

**Comment by godfreyhe** [comment]
Thanks for the review @wenlong88 @wuchong , I have update the pr based on the comments

**Review by wenlong88 — APPROVED** [review]
LGTM

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**